### ISSUE TITLE
Quit event behavior

### ISSUE THREAD

**Issue by ducharmemp** [ggez/ggez#602]
https://docs.rs/ggez/0.5.0-rc.1/src/ggez/event.rs.html#92

The default handler for keydown now calls super::quit() instead of calling the quit_event handler. Should the docs get updated or should the "run" event loop be refactored to include the keydown event again?

**Comment by ducharmemp** [comment]
I can raise a PR to do either, just want to know the direction that we should take.

**Comment by icefoxen** [comment]
That's a good question, thank you! Let's see...

The question seems to basically be a matter of, does calling `quit()` cause a quit event, or does a quit event call `quit()`? Currently it is the latter, see here: https://github.com/ggez/ggez/blob/master/src/event.rs#L161 . It looks like in 0.4 it was different, see https://github.com/ggez/ggez/blob/0.4/src/context.rs#L176 and such. Iirc the difference is there because winit doesn't let you push new events, though that may be different now; I'm on my phone so double checking is awkward.

Either way is valid, but it would be preferable to keep the 0.4 behavior the same in 0.5. If this isn't practical, we should update the docs to reflect the change.

**Comment by ducharmemp** [comment]
I think that calling the quit_event handler might be the best option, but we'd need to call out in the docs that if you override the keyboard_event handler, you're going to lose the default ability to quit on escape. I'll draft a PR later once I finish up a few things.

**Comment by icefoxen** [comment]
I mean, losing the default ability to quit on escape is intentional; the default `keyboard_event` handler is there for convenience more than anything else.  Calling it out in the docs is definitely something we want though.

**Comment by icefoxen** [comment]
Okay, so this gets ugly fast.  If we want to make `ggez::quit()` trigger a quit event, instead of a quit event calling `ggez::quit()`, here's what we have to do:

 * Shove the `EventsLoop` into the `Context`
 * Add an `Option<Event>` to the `Context`
 * Make a `poll_events()` method on the `Context` that calls the callback on the contents of that `Option` anyway, then goes through `EventsLoop::poll_events()`
 * Figure out how to make hidpi conversions work into it
 * Probably fold `Context::process_events()` into this anyway.
 * Pray that there are no weird borrowing problems.

We could just wrap the `winit::EventsLoop` in our own type, but meh.  Might have to do that though, the borrowing situation is unfortunately not trivial.

**Comment by PSteinhaus** [comment]
How about the following idea:

Instead of wrapping `winit::EventsLoop` or shoving the event loop into the Context why don't we just add a check for a new flag `ctx.quit_requested` at the end of the event loop?
- if `quit_requested` is false, then nothing happens and the loop continues normally
- if it's true (and `ctx.continuing` is still true), then a `quit_event` is triggered and its return value is checked. If it's false then `continuing` is set to false.

And `quit_requested` would simply be set to true whenever someone calls `quit` (instead of directly setting `ctx.continuing` to false, as it's done now).

The only other change necessary would be to change [these lines](https://github.com/ggez/ggez/blob/50918f133785e1ee585c022d94a0176ea0e73887/src/event.rs#L208-L212):
```rust
WindowEvent::CloseRequested => {
    if !state.quit_event(ctx) {
        quit(ctx);
    }
}
```
to
```rust
WindowEvent::CloseRequested => {
    if !state.quit_event(ctx) {
        ctx.continuing = false;
    }
}
```

**Comment by PSteinhaus** [comment]
Instead we could also let `quit` be as is and just introduce a new function `request_quit`, which then leads to a `quit_event` through setting `ctx.quit_requested` to true.

**Comment by PSteinhaus** [comment]
I implemented my proposal in 3b626da6cd6a9f5810b0e62f52a0f67929709a8c.

I'm pretty happy with it. The two open questions that I have now are:
- Is replacing `quit` with `request_quit` the right thing to do? I think so, as they have different effects/intentions, but I'm open to critique. It changes the API, though many people could probably just keep using `request_quit` like they used `quit`.
- Shouldn't we move away form module-level functions here as well? (as we did recently in all the subcontexts?) So instead of `event::request_quit` we'd have `Context::quit`. Then again it makes a bit of sense in terms of tidyness that this is placed in `event` instead of being just some function of `Context`.
